js/viz.js

```diff
@@ -60,9 +60,14 @@ const viz = {
   createCanvas(width, height) {
     const base = d3.select('#visualization');
     const canvas = base.append('canvas');
-    canvas.attr('width', width);
-    canvas.attr('height', height);
     const context = canvas.node().getContext('2d');
+    const scale = window.devicePixelRatio || 1;
+
+    canvas.attr('width', width * scale);
+    canvas.attr('height', height * scale);
+    canvas.style('width', `${width}px`);
+    canvas.style('height', `${height}px`);
+    context.scale(scale, scale);
 
```

Comment:
Following is the approach used:

Creating the Canvas element in the scaled up resolution and drawing in the context which is scaled accordingly.

The following two lines ensures that overall graph width, height is never overridden by the scaled up values.
```
canvas.style('width', `${width}px`);
canvas.style('height', `${height}px`);
```